common/metadata-helper.cpp

```diff
@@ -0,0 +1,324 @@
+#include "metadata-helper.h"
+
+#ifdef WIN32
+
+#include <Windows.h>
+#include <memory>
+#include <vector>
+#include <algorithm>
+#include <iterator>
+
+#include <regex>
+#include <iostream>
+#include <sstream>
+#include <functional>
+#include <map>
+#include <memory>
+
+#include <librealsense2/rs.hpp>
+
+#define MAX_KEY_LENGTH 255
+#define MAX_VALUE_NAME 16383
+
+#endif
+
+namespace rs2
+{
+#ifdef WIN32
+    struct device_id
+    {
+        std::string pid, mi, guid, sn;
+    };
+
+    inline std::string to_lower(std::string x)
+    {
+        transform(x.begin(), x.end(), x.begin(), tolower);
+        return x;
+    }
+
+    bool operator==(const device_id& a, const device_id& b)
+    {
+        return (to_lower(a.pid) == to_lower(b.pid) &&
+            to_lower(a.mi) == to_lower(b.mi) &&
+            to_lower(a.guid) == to_lower(b.guid) &&
+            to_lower(a.sn) == to_lower(b.sn));
+    }
+
+    class windows_metadata_helper : public metadata_helper
+    {
+    public:
+        static bool parse_device_id(std::string id, device_id* res)
+        {
+            static const std::regex regex("[\\s\\S]*pid_([0-9a-fA-F]+)&mi_([0-9]+)#[0-9]&([0-9a-fA-F]+)&[\\s\\S]*\\{([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12})\\}[\\s\\S]*");
+
+            id = to_lower(id); // ignore case
+
+            std::match_results<std::string::const_iterator> match;
+
+            if (std::regex_match(id, match, regex) && match.size() > 4)
+            {
+                res->pid = match[1];
+                res->mi = match[2];
+                res->sn = match[3];
+                res->guid = match[4];
+                return true;
+            }
+            return false;
+        }
+
+        static void do_foreach_device(std::vector<device_id> devices, 
+            std::function<void(device_id&, std::wstring)> action)
```

Comment:
It's not clear what the wstring is representing... it's a device path, right? Can we name the argument? There might be an issue with the compiler (it's ambiguous in the standard, so depends on the compiler)...

Maybe name the function 'foreach_device_path'?

The device_id should be passed as 'const &'.